### PR TITLE
Setup image upload and Blobstore

### PR DESCRIPTION
**Pull request by ananyeet**
Create a image upload page that also shows a photo gallery of all the previously uploaded images. 
image
(The same picture was uploaded 4 times, each uploaded picture is only shown once)
The title is shown as the hovertext (can't see that here because I couldn't take a screenshot)

### CODE REVIEW

**Comment by ananyeet** [comment]
#18 for the todo in ImageUploaderServlet.java.

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 1]
```diff
@@ -47,18 +47,27 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     for (Entity e : results.asIterable()) {
       String eAsJson = gson.toJson(e);
+      cmts.put(eAsJson);
       try {
         JSONObject o = new JSONObject(eAsJson);
-        comments.accumulate("comments", o);
+        cmts.put(o);
       } catch (JSONException jsonE) {
         // eAsJson will always be a json because it's generated by gson
         // so we can never actually reach this point
```
Comment (zghera) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:60]: If you never reach the catch with this exception, would it make sense to look for a different (more general exception) or just remove the try/catch block all-together?

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 2]
```diff
@@ -47,18 +47,27 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
       }
     }
 
+    try {
+      comments.put("comments", cmts);
+    } catch (JSONException jsonE) {
+      // cmts will always be a JSON because it's a JSONArray. 
+      // It's impossible to get to this line.
+    }
```
Comment (zghera) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:69]: Same comment/question as above.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
```
Comment (zghera) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:103]: Remove this comment or implement the file check. I did it by [checking the content type](https://github.com/zghera/STEP-Portfolio/blob/serve-blob-manually/portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java) but there may be another way that is better.

File: portfolio/src/main/webapp/imageupload.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  });
+}
+
+/** 
+ * Get image URL from /serve-blob.
+ */
+async function setURLFromBlob(blobKey, img) {
+  if (blobKey == "" || img == null) {
+    // Click the below link and you'll get rick rolled. 
+    return "https://images2.minutemediacdn.com/image/upload/c_fill,g_auto,h_1248,w_2220/v1555349001/shape/mentalfloss/screen_shot_2016-02-04_at_4.43.43_pm.png?itok=6RPYvfMt"
```
Comment (zghera) [inline portfolio/src/main/webapp/imageupload.js:39]: This is amazing.

File: portfolio/src/main/webapp/imageupload.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  }
+  
+  // Set up params. 
+  const params = new URLSearchParams();
+  params.append("blobKey", blobKey);
+
+  // Perform the request. 
+  // Courtesy of https://stackoverflow.com/questions/50248329/fetch-image-from-api
+  fetch("/serve-blob", {method:'POST', body:params}).then(response => response.blob()).then(image => {
```
Comment (zghera) [inline portfolio/src/main/webapp/imageupload.js:48]: Oh nice! I like this way more than my method of doing the post request to the serve-blob servlet.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    try {
+      images.put("images", imgs);
+    } catch (JSONException jsone) {
+      // imgs will always be a json because it's the string form of a JSONArray. 
+      // It's impossible to get to this line. 
+    }
+
+    return images.toString();
```
Comment (beiebi0) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:164]: Can you please clean the code up a little bit by use JsonObject and JsonArray from Gson to replace ones from JSON.
```suggestion 
  private String getImagesAsJson(){
    Gson gson = new Gson();
    JsonArray imgs = new JsonArray();

    for (Entity e : results.asIterable()) {
      imgs.put(gson.toJson(e));
    }

    JsonObject images = new JsonObject();
    images.add("images", imgs);
    return images.getAsString();
  }
```

File: portfolio/src/main/webapp/imageupload.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+/**
+ * Fetch images from blobstore. 
+ */
+function fetchImages() {
+  fetch("/img-upload", {method: "GET"}).then(response => response.json()).then(imagesJson => {
+    photoFrame = document.getElementById("uploaded-images");
+    imageList = imagesJson.images;
+    for(i of imageList) {
```
Comment (beiebi0) [inline portfolio/src/main/webapp/imageupload.js:19]: space after for ?

File: portfolio/src/main/webapp/imageupload.js (hunk 4 above)
Comment (beiebi0) [inline portfolio/src/main/webapp/imageupload.js:39]: formatting? Put it into a static variable and return here.

File: portfolio/src/main/java/com/google/sps/servlets/BlobServerServlet.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+package com.google.sps.servlets;
+
+// General imports:
+import java.io.IOException;
+// Servlet imports:
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+// Blobstore imports:
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/BlobServerServlet.java:24]: These comments are unnecessary, as the code is self-explanatory.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreUrlServlet.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+package com.google.sps.servlets;
+
+// General imports:
+import java.io.IOException;
+// Servlet imports:
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+// Blobstore imports:
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreUrlServlet.java:24]: Same as above.

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 10]
```diff
@@ -36,31 +38,34 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/comment")
 public class CommentServlet extends HttpServlet {
-    
+  private static final int MAX_COMMENTS_SERVE = 1000;
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:44]: Add method comment.

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 11]
```diff
@@ -36,31 +38,34 @@
...
+    try {
+        // create the [{comment1}, ...]
+        JSONArray commentsList = new JSONArray(new JSONTokener(gson.toJson(results)));
+       
+        commentJSON.put("comments", commentsList);
+    } catch (JSONException jsonE) {
+        // We only get here if .put's first argument is null or 
+        // its second input is not a valid JSON. 
+        // Neither of these can happen. 
+        // So this catch statement need not be implemented
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:64]: Add period.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+// JSONObject imports:
+import org.json.JSONArray; 
+import org.json.JSONObject;
+import org.json.JSONException;
+import org.json.JSONTokener;
+// GSON imports:
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:51]: All of these comments are unnecessary -- same as above.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String imagesJson = getImagesAsJson();
+    response.setContentType("application/json");
+    response.getWriter().println(imagesJson);
+  }
+
+  // Written with much assistance from examples/hello-world-fetch.
+  // (Read as: This is copied from examples/hello-world-fetch, but I had to fix some things)
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:84]: Add period.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:105]: Add period.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    // https://stackoverflow.com/q/10779564/873165
+    
+    return blobKey;
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client.
+   * Copied from TextProcessorServlet.java and modified slightly for my needs.
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:114]: Since this is private, the requirements on documentation are less strict. However, since you are including a method comment with an @return tag, you should include tags for the method parameters as well.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+  private String getParameter(HttpServletRequest request, String parameterName, String defaultValue) {
+    String result = request.getParameter(parameterName);
+    if (result == null || result.equals("") || result.length() == 0) {
+      return defaultValue;
+    }
+    return result;
+  }
+
+  private void addToDatastore(String title, BlobKey blobKey) {
+    // Filter out bad URLs
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:125]: Add period.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    if (blobKey == null) {
+      return;
+    }
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    Entity imageEntity = new Entity("Image");
+    imageEntity.setProperty("title", title);
+    imageEntity.setProperty("blobKey", blobKey);
+    imageEntity.setProperty("timestamp", System.currentTimeMillis());
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:134]: Something interesting to consider... since this timestamp is captured on the server side, all timestamps will be based on the location of the server rather than the user. This is 100% fine, I just thought it was interesting to think about and thought I should point it out.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+    datastore.put(imageEntity);
+  }
+
+  private String getImagesAsJson(){
+    Gson gson = new Gson();
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    Query q = new Query("Image").addSort("timestamp", SortDirection.ASCENDING);
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:143]: Is this "Image" the same as the "Image" on line 131? Any string, number, etc that is used in multiple places should be made a constant at the top of the file. The same goes for "timestamp" and any other strings or values referenced throughout this entire file that are duplicated.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+    List<Entity> results = ds.prepare(q).asList(FetchOptions.Builder.withLimit(MAX_IMAGES_TO_SERVE));
+    
+    // {"images": [{image 1}, {image 2} .. ]}"
+    JSONObject imagesJSON = new JSONObject(); // Contains array of "images"
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:147]: "Contains array of \"images\"" is unnecessary to state, since it's built below on line 152.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    try { 
+        // create [{image 1} ...]
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:149]: Capitalize "c" and add period.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  private String getImagesAsJson(){
+    Gson gson = new Gson();
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    Query q = new Query("Image").addSort("timestamp", SortDirection.ASCENDING);
+    List<Entity> results = ds.prepare(q).asList(FetchOptions.Builder.withLimit(MAX_IMAGES_TO_SERVE));
+    
+    // {"images": [{image 1}, {image 2} .. ]}"
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:146]: Unnecessary quote at the end of the comment?

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    JSONObject imagesJSON = new JSONObject(); // Contains array of "images"
+    try { 
+        // create [{image 1} ...]
+        JSONArray imageList = new JSONArray(new JSONTokener(gson.toJson(results)));
+        
+        imagesJSON.put("images", imageList);
+      } catch (JSONException jsone) {
+        // We only get here if .put's first argument is null or 
+        // it ssecond input is not a valid JSON.
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:155]: Fix typo "ssecond".

File: portfolio/src/main/webapp/imageupload.js [hunk 23]
```diff
@@ -0,0 +1,57 @@
+var rick_roll_link = "https://images2.minutemediacdn.com/image/upload/c_fill,g_auto,h_1248,w_2220/v1555349001/shape/mentalfloss/screen_shot_2016-02-04_at_4.43.43_pm.png?itok=6RPYvfMt"
+/**
+ * Fetch Blobstore url from /blobstore-url and link it to the form
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/imageupload.js:3]: Add period.

File: portfolio/src/main/webapp/imageupload.js [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+ */
+function fetchBlobstoreUrlAndLink() {
+  fetch('/blobstore-url').then(response => {return response.text()}).then(uploadUrl => {
+    uploadForm = document.getElementById("imageupload");
+    console.log(uploadUrl);
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/imageupload.js:8]: Remove test logging.

File: portfolio/src/main/webapp/imageupload.js [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+/** 
+ * Get image URL from /serve-blob.
+ */
+async function setURLFromBlob(blobKey, img) {
+  if (blobKey == "" || img == null) {
+    // Click the below link and you'll get rick rolled.
+    return rick_roll_link
+  }
+  
+  // Set up params.
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/imageupload.js:45]: Unnecessary comment.

File: portfolio/src/main/webapp/imageupload.js [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+  if (blobKey == "" || img == null) {
+    // Click the below link and you'll get rick rolled.
+    return rick_roll_link
+  }
+  
+  // Set up params. 
+  const params = new URLSearchParams();
+  params.append("blobKey", blobKey);
+
+  // Perform the request.
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/imageupload.js:49]: Unnecessary comment.

File: portfolio/src/main/webapp/imageupload.js [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+    // Click the below link and you'll get rick rolled.
+    return rick_roll_link
+  }
+  
+  // Set up params. 
+  const params = new URLSearchParams();
+  params.append("blobKey", blobKey);
+
+  // Perform the request. 
+  // Courtesy of https://stackoverflow.com/questions/50248329/fetch-image-from-api
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/imageupload.js:50]: Add period.

File: portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String imagesJson = getImagesAsJson();
+    response.setContentType("application/json");
+    response.getWriter().println(imagesJson);
+  }
+
+  // Written with much assistance from examples/hello-world-fetch.
+  // (Read as: This is copied from examples/hello-world-fetch, but I had to fix some things)
+  private BlobKey getUploadedUrlName(HttpServletRequest request, String formElementName) {
```
[the remark this answers is not in the document]
Comment (ananyeet) [reply portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java:79]: No, but I figured it'd be better than hardcoding it in there, just in case.

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 29]
```diff
@@ -28,15 +28,12 @@
...
-import org.json.JSONObject;
-import org.json.JSONException;
+import org.json.JSONObject; 
+import org.json.JSONException; 
 import org.json.JSONTokener;
-import com.google.gson.Gson;
-import java.util.List;
-import java.util.ArrayList;
-import java.io.IOException;
+// Servlet-related imports:
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: Please remove unnecessary comments in imports.

Comment (ananyeet) [reply portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: I removed them!! Why are they still there??

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 30]
```diff
@@ -32,6 +32,7 @@
 import org.json.JSONArray; 
 import org.json.JSONObject; 
 import org.json.JSONException; 
+import org.json.JSONTokener;
 // Servlet-related imports:
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:36]: Though not part of this change, I recommend you remove these import comments in another PR.
